Reorganize the explainer a bit (#73)
davidben/tls-trust-expressions: explainer.md

## Introduction

> While they differ in approach, both drafts aim to solve the same problem: **enable [TLS](https://www.rfc-editor.org/rfc/rfc8446) implementations to efficiently authenticate to peers with diverse [certificate](https://www.rfc-editor.org/rfc/rfc5280) policies.** This avoids a conflict between service availability and user security.

Reviewer: policies here sounds a bit too fluffy and vague for the problem statement - can we stick to divers sets of trust anchors or something like that (Trust anchors can be referenced straight out of 5280 I believe)

> While they differ in approach, both drafts aim to solve the same problem: **enable [TLS](https://www.rfc-editor.org/rfc/rfc8446) implementations to efficiently authenticate to peers with diverse [certificate](https://www.rfc-editor.org/rfc/rfc5280) policies.** This avoids a conflict between service availability and user security.

Reviewer 2: This sort of reads like a requisite lead-in paragraph was deleted, and if you squint, this still reads like the goal of a solution more than a problem statement. I think the conflict between availability and security might better be described as a corollary of the below problem statement, but I'm open to iterating over this if it's headed in the right direction.

==========
Background:
Today, TLS subscribers are provisioned with one or more certification paths which they use to authenticate themselves to TLS relying parties. Currently, TLS subscribers lack a reliable way of knowing which among these provisioned paths will be accepted by a given relying party, if any. Absent a clear signal, TLS subscribers are forced to guess, which causes connection failures when this guess is incorrect.

In the scenario in which a TLS subscriber wishes to authenticate itself to relying parties with distinct or mutually-exclusive trust settings, there is no single correct answer. Advanced TLS implementations may attempt to use TLS fingerprinting to make informed guesses; however, this is unreliable and client trust may change over time, causing existing fingerprinting mechanisms and the subsequent certification path selection to fail. 

Problem:
Trust Anchor IDs and Trust Expressions present two different approaches to solve the same problem: eliminating the uncertainty around what relying parties trust so that TLS subscribers can obtain, and intelligently serve, a trusted certification path to TLS relying parties.

Reviewer: I generally like this, other than some degree of discomfort between "today subscribers are provisioned witth one or more" - I think we risk the reader believing today it isn't the case and most of the time it's only one.. yeah I know you say one or more, but I think a lot of our readership is stuck in a mode of "who cares what we have today works fine"

Reviewer 2: "1 or more" is necessary for the fingerprinting --> cert selection to make sense. Perhaps we can more clearly emphasize that guesses and resulting breakage do happen, if that assuages your concerns about people thinking this doesn't actually happen today (which it does).

Author: I wanted to try a consistent framing between "this is make `certificate_authorities` more useful" and "this is straight-up impossible". I think the former is a better framing because a lot of deployments *can* use `certificate_authorities` today. Like it's just fine for client certs.

How's this new version?

We refer to the authenticating party as the *subscriber* and the peer as the *relying party*. The common case in TLS is server certificate authentication, where the subscriber is the server, and the relying party is the client. The roles are reversed with client certificates. For clarity, this document will primarily discuss the server certificate case, but most of the motivations and solutions apply analogously to client certificates.

> To solve this problem, the two drafts extend the existing TLS 1.3 [`certificate_authorities` extension](https://www.rfc-editor.org/rfc/rfc8446#section-4.2.4) to reduce the size cost, making it usable for more applications. They also define a supporting ACME extension to help servers provision multiple certificates. As with other internet drafts at this stage of the process, they are not final mechanisms. Rather, they are starting points, intended to demonstrate feasibility and some approaches, for the working group to build on for a complete solution.

Reviewer 2: Suggest: "extend the pattern/mechanism used by the existing TLS 1.3 `certificate_authorities` extension"

Author: Done.

> To solve this problem, the two drafts extend the existing TLS 1.3 [`certificate_authorities` extension](https://www.rfc-editor.org/rfc/rfc8446#section-4.2.4) to reduce the size cost, making it usable for more applications. They also define a supporting ACME extension to help servers provision multiple certificates. As with other internet drafts at this stage of the process, they are not final mechanisms. Rather, they are starting points, intended to demonstrate feasibility and some approaches, for the working group to build on for a complete solution.

Reviewer 2: nit: including 'extension' in the link after the codeblock formatting causes a weird break that makes it look like it's two separate links. Maybe just linkify `certificate_authorities`

Author: Done.

> The remainder of this section discusses what this problem statement means, and why it is important to solve. Subsequent sections provide an overview of the two solutions, discussion on design goals and key cases, and finally some alternatives we considered, including various existing solutions (e.g. fingerprinting and cross-signing), which do not meet the requirements.

Reviewer: ```suggestion
The remainder of this section discusses what this problem statement means, and why it is important to solve. Subsequent sections provide an overview of the two solutions, discussion on design goals and key cases, and finally some alternatives we considered, including various existing solutions (e.g. fingerprinting and cross-signing), which do not meet the design goals.
```

### Background

TLS uses X.509 certificates to associate DNS names, or other application identifiers, with the authoritative server's TLS keys. These associations are signed by certificate authorities (CAs) and are presented by servers to clients. Each client curates a set of CAs, called *trust anchors*, whose associations the client accepts. If the client's trust anchors can be trusted to only issue correct associations, the client can rely on TLS to securely connect to servers.

In this system, the client's trust anchors directly impact service availability and user security:

> * If the authoritative server is unable to obtain or present a certificate from a CA trusted by the client, it cannot serve that client. Without trustworthy CAs that servers can use, the client cannot connect.

Reviewer: I really think the use of "CA" interchanably with "key" hurts us in this section.. I know it is sort of the "standard" in these documents and I hate it.

Author: We're not using it interchangeably.

A CA is a kind of service that, among other things, uses a key. That service is defined by far more than a key. There's a name and other properties, usually packaged in a self-signed certificate. There's also the software and processes that access and use that key.

A CA operator is an entity who operates one or more CAs.

The client contains a list of CAs, not CA operators. Rarely but sometimes CAs change hands between CA operators, though that is quite a process. When clients make trust decisions, they certainly take CA operators into account, but the trust anchor list is ultimately a list of CAs.

We've been pretty consistent throughout all our documents here. I think this is correct. Saying "key" loses all the other stuff that is part of a CA.

Reviewer: Not the hill to die on today then, but really, I find using the language "untrustworthy" referring to a "CA" as opposed to a "CA operator" a bit jarring.

This dynamic means PKIs necessarily evolve over time. To meet user security needs, clients remove compromised or untrustworthy CAs, so that they cannot be used to intercept TLS. To meet service availability needs, clients add trustworthy CAs that add net value to the ecosystem, so that servers are able to obtain acceptable certificates.

Some transitions combine the two: a long-lived CA key is more likely to be compromised and presents a higher user security risk. To rotate a key in a PKI, the client would add a new key from the CA operator, and then remove the old one.

### Client Diversity

As with other TLS parameters, client populations do not have completely uniform trust anchors.

PKI transitions described above inherently create this diversity, if for no other reason than out-of-date clients in the ecosystem. These clients could range from unupdatable TV set-top boxes to some IoT device to an individual user's browser that has not yet, or could not, communicate with its update service.

> Beyond temporal divergence, different clients have different needs and make different decisions on behalf of their users. For example, within the Web PKI, root programs make trust decisions independently. Some more specialized clients, such as mobile apps, may even pin to one or two specific CAs.

Reviewer 2: We can avoid people rabbit-holing on pinning by saying something more generally true, such as:

"Some more specialized clients, such as mobile apps, may be constrained to trusting one or two specific CAs"

Author: Applied.

To support a potentially diverse set of clients, the server must present an acceptable certificate to each one. TLS has a standard solution for this across its many parameters: negotiation. For client certificates, the CertificateRequest message listed trusted CAs since SSL 3.0 and TLS 1.0. TLS 1.3 generalized this to server certificates with the [`certificate_authorities` extension](https://www.rfc-editor.org/rfc/rfc8446#section-4.2.4), allowing [clients](https://www.rfc-editor.org/rfc/rfc8446#section-4.4.2.3) and [servers](https://www.rfc-editor.org/rfc/rfc8446#section-4.4.2.2) to select between available certificates by the peer's trust anchors.

However, `certificate_authorities`'s size is impractical for some applications. Existing PKIs may have many CAs, and existing CAs may have long X.509 names. As of August 2023, the [Mozilla CA Certificate Program](https://wiki.mozilla.org/CA/Included_Certificates) contained 144 CAs, with an average name length of around 100 bytes. Such TLS deployments often do not use trust anchor negotiation at all.

### Security vs Availability

> Without a negotiation mechanism, service availability and user security conflict, with user security paying much of the cost.

Reviewer 2: Suggest something like: "Without a negotiation mechanism, changes to client trust to improve user security directly conflict with service availability, with user security being deprioritized in favor of avoiding any kind of breakage"

IMO, framing user security as the cause of client diversity makes this point more clearly.

Author: Applied.

To support this compression scheme, participating CAs and root programs must coordinate to configure consistent information in clients and servers. To do this, the CA periodically fetches a "trust store manifest" from the root program. The protocol additionally must solve a "version skew" problem where the client references a newer trust store version than the server knows of. Most of the protocol's complexity comes from these two parts.

### Trust Anchor Identifiers

Trust Anchor Identifiers consists of three parts:

1. To mitigate large X.509 names, short "trust anchor identifiers" uniquely identify each participating CA, at an estimated 5 bytes per CA. These are sent in a `trust_anchors` extension, analogous to `certificate_authorities`.

2. Bandwidth and privacy constraints may still prevent a client from enumerating its full CA list. Server use an extension to [HTTPS/SVCB](https://www.rfc-editor.org/rfc/rfc9460.html) to publish available trust anchors in DNS. The client fetches this, selects the desired option, and requests it in `trust_anchors`.

> 3. To accommodate stale DNS records, or cases where the DNS mechanism does not work, servers also send available trust anchors in the TLS handshake. If the client rejects the certificate, it can use this to retry with a more accurate `trust_anchors` request.

Reviewer: ```suggestion
3. To accommodate stale or missing DNS records, servers also send available trust anchors in the TLS handshake. If the client rejects the certificate, it can use this to retry with a more accurate `trust_anchors` request.
```

Author: Missing is weird because if it's missing the server should just deploy them. But I do like the tighter phrasing. I did "stale or unavailable".

* Trust Expressions require that CAs continually fetch information from manifests that are published by root programs, while Trust Anchor IDs rely only on static pre-assigned trust anchor identifiers.

* Trust Anchor IDs, when trust anchors are conditionally sent, have different fingerprinting properties. See [Privacy Considerations](https://davidben.github.io/tls-trust-expressions/draft-beck-tls-trust-anchor-ids.html#name-privacy-considerations).

* Trust Anchor IDs can only express large client trust stores (for server certificates), not large server trust stores. Large trust stores rely on the retry mechanism described in [the draft](https://davidben.github.io/tls-trust-expressions/draft-beck-tls-trust-anchor-ids.html#name-retry-mechanism), which is not available to client certificates.

The two mechanisms can be deployed together. A subscriber can have metadata for both mechanisms available, and a relying party can advertise both. Mechanisms with aspects of both may potentially also be designed.

### Server Configuration

> Both drafts anticipate a model where servers provision a collection of candidate certificate paths, with some associated selection metadata, instead of a single path. When a client connect, the server software automatically selects one to send by matching the ClientHello message against the selection metadata, along with other TLS criteria such as ECDSA vs RSA. If there are multiple matches, server software chooses based on its own criteria, such as certificate size.

Reviewer: ```suggestion
Both drafts anticipate a model where servers automatically provision a collection of candidate certificate paths, with some associated selection metadata, instead of a single path. When a client connect, the server software automatically selects one to send by matching the ClientHello message against the selection metadata, along with other TLS criteria such as ECDSA vs RSA. If there are multiple matches, server software chooses based on its own criteria, such as certificate size.
```

Author: I don't think the provisioning is necessarily automatic. E.g. your backup certificate idea. Automation is very valuable, but that's covered in the next section.

> Both drafts anticipate a model where servers provision a collection of candidate certificate paths, with some associated selection metadata, instead of a single path. When a client connect, the server software automatically selects one to send by matching the ClientHello message against the selection metadata, along with other TLS criteria such as ECDSA vs RSA. If there are multiple matches, server software chooses based on its own criteria, such as certificate size.

Reviewer 2: nit: make sure we're consistent about "certification paths" vs. "certificate paths"

Author: For better or worse, this document uses "certificate paths" everywhere. I don't care either way which one we use but matched for now. It's all kind of a mess. Colloquial terminology is overwhelming "certificate path" and "certificate authority". 5280 says "certification path" and "certification authority". 8446 says "certification path" but "certificate authority".

> If the negotiation mechanism is either not supported or did not match a candidate, servers should behave as they do today, possibly falling back to a single default certificate or heuristics. However, each client that deploys negotiation no longer constrains that fallback path, reducing the burden the server operator in configuring it.

Reviewer: ```suggestion
If the negotiation mechanism is either not supported or did not match a candidate, servers should behave as they do today, possibly falling back to a single default certificate or heuristics. However, each client that deploys negotiation no longer constrains that fallback path, increasing the choices available for a valid fallback certificate for the remaining clients.
```

## Design Goals

In solving this problem, there were a few notable design goals to meet:

* Enable PKIs to evolve as needed for user security, in a timely manner and without conflicting with availability. We are particularly focused on PKIs that share characteristics with the existing Web PKI.
* Minimize bandwidth cost in the TLS handshake. With the post-quantum transition, cryptographic material will become [much more expensive](https://dadrian.io/blog/posts/pqc-signatures-2024/) to send.
* Minimize burden to server operators, particularly avoiding ongoing manual work. Most deployments should leave ongoing decisions to TLS software, ACME software, and ACME server behavior.

We discuss these in more detail below:

### Timely PKI Evolution

If a CA is compromised or otherwise considered untrustworthy, mitigating the TLS interception attacks requires that the client no longer trust the CA. This means we should minimize delays in completing PKI transitions at the client, as that delay translates to time during which users remain at risk.

> In other transitions, reducing delays is also valuable. If trustworthy CA is considered to provide net value to the ecosystem and added, a timely transition allows servers to make use of the CA and benefit.

Reviewer: ```suggestion
In other transitions, reducing delays is also valuable. If trustworthy CA is considered to provide net value to the users of a client and is added, a timely transition allows servers to make use of the CA and provide the benefit of using it to the client.
```

Author: "net value to the ecosystem" is the phrasing we used before. I think that is a better phrasing than to a client.

There are many reasons to add CAs. A CA might provide better automation, which directly benefits the server operator. Of course, it indirectly benefits the user by making the system more robust, which is ultimately what this is for. But this is a complicated and broad, so "ecosystem" is more appropriate.

I rephrased a bit so we don't talk about the server benefiting because that's also poor phrasing for the same reason. Also added a couple examples.

> Negotiation mechanisms naturally achieve this, following the standard TLS pattern, by letting the server serve any collection of certificates it needs to maintain availability. The server is not constrained by availability of CA cross-signs, which carry signficant risks for all parties in the PKI.

Reviewer: What risks? (this is the question that will be asked, do we leave it unanswered here?)

Reviewer 2: It feels like we can TODO a link to separate text that explains this later.

Author: I mean, the separate text *already exists*. It's Considered Alternatives. I rewrote it slightly and linked there.

The ACME multi-certificate extensions can further reduce user security risk, by allowing ACME servers to automatically manage some transitions. For example, when rotating CA keys, the ACME server can automatically provision chains for both the old and new key.

### Minimizing Bandwidth Cost

We wish to minimize bandwidth cost in the TLS handshake. This includes bandwidth from the negotiation mechanism itself, and bandwidth from the certificates that are negotiated. Increased handshake bandwidth leads to delays in serving users (e.g. slower webpage loads) and consumes resources like the user's data plan.

> The two drafts reduce the bandwidth costs of the `certificate_authorities` extension. For PKIs where `certificate_authorities` was already viable, this is a size optimization over the existing deployment model. For PKIs where `certificate_authorities` was prohibitively large, this provides agility with minimal size cost.

Reviewer: ```suggestion
The two drafts reduce the bandwidth costs of the `certificate_authorities` extension. For PKIs where `certificate_authorities` was already viable, this is a size optimization over the existing deployment model. For PKIs where `certificate_authorities` was prohibitively large, this provides negotiaiton with minimal size cost.
```

Author: Applied but with spelling fix.
